clippy_lints/src/methods/iter_overeager_cloned.rs

```diff
@@ -35,10 +58,33 @@ pub(super) fn check<'tcx>(
             return;
         }
 
-        let (lint, msg, trailing_clone) = if is_count {
-            (REDUNDANT_CLONE, "unneeded cloning of iterator items", "")
-        } else {
-            (ITER_OVEREAGER_CLONED, "unnecessarily eager cloning of iterator items", ".cloned()")
+        if let Op::NeedlessMove(_, expr) = op {
+            if let rustc_hir::ExprKind::Closure(closure) = expr.kind
+               && let Body {params: [p], value: _, generator_kind: _ } = cx.tcx.hir().body(closure.body) {
+                    match p.pat.kind {
+                        rustc_hir::PatKind::Binding(BindingAnnotation(_, Mutability::Not), _, _, _) => {
+                            let body = cx.tcx.optimized_mir(closure.def_id);
+                            if let VarDebugInfoContents::Place(p) = &body.var_debug_info[0].value {
+                                let mut find_move = FindMove::new(p.local);
+                                find_move.visit_body(body);
+                                if find_move.found {
```

Comment:
If I'm understanding this correctly, this is just checking if it returns something, right? Since otherwise, if it doesn't return a `()`, it'll always have a move/copy, no matter what. If so, then why aren't you checking the return type of the closure?

If you aren't, then I'm really *really* confused on what this does. Maybe you meant to skip `Copy`s? (If so, you probably want to use `ExprUseVisitor` and overwrite `consume` instead of `optimized_mir`.)

Reply:
I want to check whether the parameter of closure is `move`d in body, `_2`.
I'm not sure a better way to do that.

If it's `Copy`, it think should be checked in line 52:
```
&& matches!(*iter_assoc_ty.kind(), ty::Ref(_, ty, _) if !is_copy(cx, ty))
```

Reply:
> I want to check whether the parameter of closure is moved in body, _2.

Wow, I completely missed this is `var_debug_info`...

This, however, doesn't always work. For example, `_ = s` generates absolutely zero code, no matter what. You're basically relying on MIR not improving, ever, and if it does you're screwed.

MIR is generally not the way to do this. It loses a lot of information and fluff that the HIR has, which is necessary for linting.

> I'm not sure a better way to do that.

Why aren't you using a single `for_each_expr`? You could just check if there are any paths referencing the local.

Reply:
I'm afraid `for_each_expr` can't detect `move` .
In hir, we can't detect whether `x.len()` use `&self` or `Self` ?

Reply:
> In hir

We can, using the `ExprUseVisitor` and overwriting `consume`.

`&self` and `Self` are both `move` so in MIR, afaik, you're linting both. For `mir-opt-level` > 0, it's lowered to a `Len`, but above that it's a call and a `move`.